core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java

```diff
@@ -38,4 +55,98 @@ public void deleteFile(String path) {
       throw new RuntimeIOException(e, "Failed to delete file: %s", path);
     }
   }
+
+  @Override
+  public InputFile readMetadataFile(String fileName) {
+    return newInputFile(metadataPath(fileName).toString());
+  }
+
+  @Override
+  public OutputFile newMetadataFile(String fileName) {
+    return newOutputFile(metadataPath(fileName).toString());
+  }
+
+  @Override
+  public OutputFile newPartitionedDataFile(
+      PartitionSpec partitionSpec, StructLike filePartition, String fileName) {
+    String location;
+    if (useObjectStorage) {
+      // try to get db and table portions of the path for context in the object store
+      String context = pathContext(new Path(newDataFileLocation));
+      String partitionAndFilename = String.format(
+          "%s/%s", partitionSpec.partitionToPath(filePartition), fileName);
+      int hash = HASH_FUNC.apply(partitionAndFilename);
+      location = String.format(
+          "%s/%08x/%s/%s/%s",
+          objectStorePath,
+          hash,
+          context,
+          partitionSpec.partitionToPath(filePartition),
+          fileName);
+    } else {
+      location = String.format(
+          "%s/%s/%s",
+          newDataFileLocation,
+          partitionSpec.partitionToPath(filePartition),
+          fileName);
+    }
+    return newOutputFile(location);
+  }
+
+  @Override
+  public OutputFile newUnpartitionedDataFile(String fileName) {
+    return newOutputFile(String.format("%s/%s", newDataFileLocation, fileName));
+  }
+
+  private Path metadataPath(String filename) {
+    return new Path(new Path(tableLocation, "metadata"), filename);
+  }
+
+  public void updateProperties(Map<String, String> newTableProperties) {
```

Comment:
This is a concession to the fact that when table properties are updated, it's possible that the data path could change. It's not immediately clear if we can do better than this.

Reply:
Actually `Supplier` isn't great because the `Supplier` itself isn't serializable. The `Supplier` would be tied to the `HadoopTableOperations` object.

Reply:
That makes sense. I think the best option is what you originally had, but with `TableMetadata`, `updateTableMetadata`. That's `Serializable` right?

Reply:
You could also make `tableMetadataSupplier` transient and cache the result. Then at least the FileIO pulls metadata when needed instead of requiring a push.

Reply:
Okay, so `HadoopFileIO` would need to extract paths from `TableMetadata` each time `updateMetadata` is called instead of storing a reference. I think it is still better to pass the whole `TableMetadata` in case anything else changes. Probably also note in javadoc that `TableMetadata` isn't serializable, so don't save a copy of it.

Reply:
`updateMetadata` is a fitting compromise here!